### PR TITLE
feat(core): ship schematics schemas as public api

### LINKED ISSUE #2310: ship/expose public packages schematics schema definitions
## Feature Request

When using  external schematic from public `@nrwl` packages, there is no way to how to acquire what kind of schema props are available (except going to the docs(which are not always up to date) or going to github source code...).

I'd be helpful to ship schema definitions for all public facing nrwl schematics as part of their public api.

So usage would  look like following:  

```ts
import { LibSchema } from '@nrwl/workspace'
import { CustomSchema } from './schema'

export default function(schema: CustomSchema): Rule {
  return chain([
    // proper type checking and DX 🙏
    externalSchematic<LibSchema & Schema>('@nrwl/workspace', 'lib', {
      name: schema.name,
      tags: getTags(schema),
    }),
  ]);
}
```

### PR DESCRIPTION
**Pull request by Hotell**
## Related Issue(s)

- this is also exposing usage of TS globals and fixes those issues
 
Fixes #2310

### CODE REVIEW

File: packages/react/index.ts [hunk 1]
```diff
@@ -1,3 +1,9 @@
 export { extraEslintDependencies, reactEslintJson } from './src/utils/lint';
 export { CSS_IN_JS_DEPENDENCIES } from './src/utils/styled';
 export { assertValidStyle } from './src/utils/assertion';
+
+export { Schema as LibrarySchema } from './src/schematics/library/schema';
+export { Schema as ApplicationSchema } from './src/schematics/application/schema';
+export { Schema as ComponentSchema } from './src/schematics/component/schema';
+export { Schema as ReduxSchema } from './src/schematics/redux/schema';
+export { StorybookConfigureSchema } from './src/schematics/storybook-configuration/schema';
```
Comment (juristr) [inline packages/react/index.ts:9]: minor detail, but would rename to `StorybookConfigurationSchema` :)

**Comment by juristr** [comment]
I like the idea 👍 . Gonna ping other ppl on our internal slack to have them have a look at this as well

File: packages/react/index.ts (hunk 1 above)
Comment (Hotell) [reply packages/react/index.ts:9]: yup, that's the plan after #3455 will be merged

**Comment by vsavkin** [comment]
I like it. Quick question: why not to have types called say ApplicationSchema? Why call them Schema and then reexport it as ApplicationSchema?

**Comment by Hotell** [comment]
> why not to have types called say ApplicationSchema

we can do it, no problem, although here are my thoughts:

I'd propose to generate those interface from `schema.json` automatically, otherwise they might be and are, getting out of sync. With that in place, the interface name would be picked from `id` name within schema 👌

WDYT  folks?

**Comment by juristr** [comment]
> I'd propose to generate those interface from schema.json automatically, otherwise they might be and are, getting out of sync. With that in place, the interface name would be picked from id name within schema 👌

Yeah I thought about that as well (having a note somewhere on it). I know the Angular CLI team does it and honestly that would help make sure they are always in sync. What do you think @vsavkin

**Comment by vsavkin** [comment]
We don't want to use any utilities provided by the Angular CLI team cause we are moving away from depending on that stack as much as we can. If there is a generic generation tool, we can assess it and perphaps use it. If not, I'd not author a new one.

The tradeoff here is:
1. The cost of keeping the interfaces up to date
2. The cost of authoring the tool + the cost of integrating the tool

I think we tend to overesitamte the cost of (1) and underestimate the cost of (2) :)

Regardless of the tool though, I think the change is good. So once you resolve the conflicts I will merge it.

One thing I'd like to ask is you move normalized schemas into schematic files, so schema.ts will always have one token called Schema and nothing else.

**Comment by vsavkin** [comment]
One minor comment. The rest looks good.

**Comment by vsavkin** [comment]
@Hotell any update on this?

**Comment by vsavkin** [comment]
@Hotell I'm really sorry, but I'm going to close this one. I'm trying to keep the list of PR somewhat up to date. But if you reopen it, I'll take a look right away.

**Pull request closed without merging**